src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderValue.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Formats.Cbor;
+
+namespace System.Security.Cryptography.Cose
+{
+    public readonly struct CoseHeaderValue : IEquatable<CoseHeaderValue>
+    {
+        public readonly ReadOnlyMemory<byte> EncodedValue { get; }
+
+        private CoseHeaderValue(ReadOnlyMemory<byte> encodedValue)
+        {
+            EncodedValue = encodedValue;
+        }
+
+        public static CoseHeaderValue FromEncodedValue(ReadOnlySpan<byte> encodedValue)
+        {
+            // We don't validate here as we need to know in which label the value is going to be used to validate even more semantics.
+            var encodedValueCopy = new ReadOnlyMemory<byte>(encodedValue.ToArray());
+            CoseHeaderValue value = new CoseHeaderValue(encodedValueCopy);
+            return value;
+        }
+
+        public static CoseHeaderValue FromEncodedValue(byte[] encodedValue)
+            => FromEncodedValue(encodedValue.AsSpan());
+
+        public static CoseHeaderValue FromInt32(int value)
+        {
+            var writer = new CborWriter();
+            writer.WriteInt32(value);
+
+            return FromEncodedValue(writer.Encode());
+        }
+
+        public static CoseHeaderValue FromString(string value)
+        {
+            var writer = new CborWriter();
+            writer.WriteTextString(value);
+
+            return FromEncodedValue(writer.Encode());
+        }
+
+        public static CoseHeaderValue FromBytes(ReadOnlySpan<byte> value)
+        {
+            var writer = new CborWriter();
+            writer.WriteByteString(value);
+
+            return FromEncodedValue(writer.Encode());
+        }
+
+        public static CoseHeaderValue FromBytes(byte[] value)
+        {
+            var writer = new CborWriter();
+            writer.WriteByteString(value);
+
+            return FromEncodedValue(writer.Encode());
+        }
+
+        public int GetValueAsInt32()
+        {
+            var reader = new CborReader(EncodedValue);
+            int retVal = reader.ReadInt32();
+
+            if (reader.BytesRemaining != 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapCborEncodedValueNotValid));
+            }
+
+            return retVal;
+        }
```

Comment:
What are the exceptions for this method, and what do you want them to be?

If the value is a string, I think it's a CborContentException.  If it's an integral type but doesn't fit in 32 bits that's apparently an InvalidOperationException (reporting that the reader isn't in a state where that method made sense), and if it has trailing data then it's an InvalidOperationException from here.

Since we've kept System.Formats.Cbor out of the API for this library, we probably don't want CborContentExceptions to be leaking out.  InvalidOperationException does make sense here for "I can't give you this value as an Int32", but I think we should try/catch/wrap the CborContentException.

Reply:
> If it's an integral type but doesn't fit in 32 bits that's apparently an InvalidOperationException

That is acutally an OverflowException.

Reply:
Wraping the exceptions you listed + the overflow one seems like the best to do here.